File: docs/js/ai/pathfinding.js
import { CONFIG } from '../config.js';
import { getState } from '../state.js';
import { gridIndex, isWall } from '../grid.js';

// Pre-defined directions: [dc, dr, wallIndex]
const DIRECTIONS = [[0, -1, 0], [1, 0, 1], [0, 1, 2], [-1, 0, 3]];

/**
 * Simple binary heap for A* priority queue
 * O(log n) push and pop operations instead of O(n) splice
 */
class MinHeap {
  constructor() {
    this.heap = [];
  }

  push(node, priority) {
    this.heap.push({ node, priority });
    this._bubbleUp(this.heap.length - 1);
  }

  pop() {
    if (this.heap.length === 0) return null;
    const min = this.heap[0];
    const last = this.heap.pop();
    if (this.heap.length > 0) {
      this.heap[0] = last;
      this._bubbleDown(0);
    }
    return min.node;
  }

  isEmpty() {
    return this.heap.length === 0;
  }

  _bubbleUp(i) {
    while (i > 0) {
      const parent = Math.floor((i - 1) / 2);
      if (this.heap[parent].priority <= this.heap[i].priority) break;
      [this.heap[parent], this.heap[i]] = [this.heap[i], this.heap[parent]];
      i = parent;
    }
  }

  _bubbleDown(i) {
    const len = this.heap.length;
    while (true) {
      const left = 2 * i + 1;
      const right = 2 * i + 2;
      let smallest = i;
      if (left < len && this.heap[left].priority < this.heap[smallest].priority) {
        smallest = left;
      }
      if (right < len && this.heap[right].priority < this.heap[smallest].priority) {
        smallest = right;
      }
      if (smallest === i) break;
      [this.heap[smallest], this.heap[i]] = [this.heap[i], this.heap[smallest]];
      i = smallest;
    }
  }
}

/**
 * Find path from player to target using A* algorithm with Manhattan heuristic
 * Uses binary heap priority queue for O(log n) operations
 * @param {Object} fromPlayer - Player object with x, y, size properties
 * @param {number} targetX - Target X coordinate in pixels
 * @param {number} targetY - Target Y coordinate in pixels
 * @returns {Array} Array of cell objects forming the path, or empty if no path
 */
export function findPathToTarget(fromPlayer, targetX, targetY) {
  const state = getState();
  if (!state.maze || state.maze.length === 0) return [];

  let startC = Math.floor((fromPlayer.x - CONFIG.MAZE_OFFSET_X + (fromPlayer.size / 2)) / CONFIG.CELL_SIZE);
  let startR = Math.floor((fromPlayer.y + (fromPlayer.size / 2)) / CONFIG.CELL_SIZE);
  let endC = Math.floor((targetX - CONFIG.MAZE_OFFSET_X) / CONFIG.CELL_SIZE);
  let endR = Math.floor(targetY / CONFIG.CELL_SIZE);

  startC = Math.max(0, Math.min(startC, CONFIG.COLS - 1));
  startR = Math.max(0, Math.min(startR, CONFIG.ROWS - 1));
  endC = Math.max(0, Math.min(endC, CONFIG.COLS - 1));
  endR = Math.max(0, Math.min(endR, CONFIG.ROWS - 1));

  const start = gridIndex(startC, startR);
  const end = gridIndex(endC, endR);

  if (!start || !end) return [];

  // Reset A* state
  for (let i = 0; i < state.maze.length; i++) {
    state.maze[i].gCost = Infinity;
    state.maze[i].parent = null;
  }

  // Manhattan distance heuristic
  const heuristic = (c, r) => Math.abs(c - endC) + Math.abs(r - endR);

  const heap = new MinHeap();
  start.gCost = 0;
  heap.push(start, heuristic(startC, startR));

  while (!heap.isEmpty()) {
    const curr = heap.pop();

    // Found target
    if (curr === end) break;

    // Skip if we've already found a better path to this node
    // (can happen with duplicate entries in heap)
    if (curr.gCost === Infinity) continue;

    for (let i = 0; i < 4; i++) {
      const d = DIRECTIONS[i];
      const nc = curr.c + d[0];
      const nr = curr.r + d[1];
      const neighbor = gridIndex(nc, nr);

      // Check if neighbor is valid and reachable (no wall blocking)
      if (!neighbor || curr.walls[d[2]] || neighbor.walls[(d[2] + 2) % 4]) continue;

      const newG = curr.gCost + 1;
      if (newG < neighbor.gCost) {
        neighbor.gCost = newG;
        neighbor.parent = curr;
        const f = newG + heuristic(nc, nr);
        heap.push(neighbor, f);
      }
    }
  }

  // No path found
  if (end.gCost === Infinity) return [];

  // Reconstruct path
  const path = [];
  let temp = end;
  while (temp) {
    path.push(temp);
    temp = temp.parent;
  }
  return path.reverse();
}

export function isPlayerStuck(player) {
  if (!player.lastPos) return false;
  let dx = Math.abs(player.x - player.lastPos.x);
  let dy = Math.abs(player.y - player.lastPos.y);
  return (dx < 0.3 && dy < 0.3);
}

/**
 * Get a direction to escape when stuck
 * Improved version checks walls and prefers directions opposite to last movement
 * @param {Object} player - Player object with position and lastDir (optional)
 * @returns {{x: number, y: number}} Direction vector to move
 */
export function getUnstuckDirection(player = null) {
  const directions = [
    { x: 1, y: 0 }, { x: -1, y: 0 }, { x: 0, y: 1 }, { x: 0, y: -1 },
    { x: 1, y: 1 }, { x: -1, y: 1 }, { x: 1, y: -1 }, { x: -1, y: -1 }
  ];

  // If no player position, fall back to random shuffle
  if (!player) {
    for (let i = directions.length - 1; i > 0; i--) {
      let j = Math.floor(Math.random() * (i + 1));
      [directions[i], directions[j]] = [directions[j], directions[i]];
    }
    return directions[0];
  }

  const checkDist = 3; // pixels to check ahead
  const playerCenterX = player.x + (player.size || 3) / 2;
  const playerCenterY = player.y + (player.size || 3) / 2;

  // Filter directions to only wall-free options
  const validDirections = directions.filter(dir => {
    const checkX = playerCenterX + dir.x * checkDist;
    const checkY = playerCenterY + dir.y * checkDist;
    return !isWall(checkX, checkY);
  });

  // If no valid directions, fall back to random (we're trapped)
  if (validDirections.length === 0) {
    return directions[Math.floor(Math.random() * directions.length)];
  }

  // If player has last direction, prefer opposite direction
  if (player.lastDir && (player.lastDir.x !== 0 || player.lastDir.y !== 0)) {
    const lastDirX = player.lastDir.x > 0 ? 1 : (player.lastDir.x < 0 ? -1 : 0);
    const lastDirY = player.lastDir.y > 0 ? 1 : (player.lastDir.y < 0 ? -1 : 0);

    // Score directions: prefer opposite to last movement
    const scored = validDirections.map(dir => {
      let score = 0;
      // Opposite direction bonus
      if (dir.x === -lastDirX && lastDirX !== 0) score += 2;
      if (dir.y === -lastDirY && lastDirY !== 0) score += 2;
      // Perpendicular is also good
      if (dir.x !== 0 && lastDirX === 0) score += 1;
      if (dir.y !== 0 && lastDirY === 0) score += 1;
      // Small random factor to avoid predictability
      score += Math.random() * 0.5;
      return { dir, score };
    });

    scored.sort((a, b) => b.score - a.score);
    return scored[0].dir;
  }

  // No last direction, pick random from valid
  return validDirections[Math.floor(Math.random() * validDirections.length)];
}

File: pi/ai/pathfinding.py
# pi/ai/pathfinding.py
# A* pathfinding for AI, ported from docs/js/ai/pathfinding.js

import math
import random

from config import COLS, ROWS, CELL_SIZE, MAZE_OFFSET_X, BASE_SPEED
from grid import grid_index, is_wall

# Pre-defined directions: (dc, dr, wall_index)
_DIRECTIONS = [(0, -1, 0), (1, 0, 1), (0, 1, 2), (-1, 0, 3)]


# ---------------------------------------------------------------------------
# MinHeap — binary heap priority queue (used by A*)
# ---------------------------------------------------------------------------

class MinHeap:
    """
    Binary min-heap.  push(priority, item) / pop() → (priority, item).
    Mirrors the JS MinHeap but uses (priority, item) tuples.
    """

    def __init__(self):
        self._heap = []  # list of [priority, item]

    def push(self, priority, item):
        self._heap.append([priority, item])
        self._bubble_up(len(self._heap) - 1)

    def pop(self):
        if not self._heap:
            return None
        # Swap root with last element, remove last, then sift down
        min_item = self._heap[0]
        last = self._heap.pop()
        if self._heap:
            self._heap[0] = last
            self._bubble_down(0)
        return (min_item[0], min_item[1])

    def is_empty(self):
        return len(self._heap) == 0

    def _bubble_up(self, i):
        while i > 0:
            parent = (i - 1) // 2
            if self._heap[parent][0] <= self._heap[i][0]:
                break
            self._heap[parent], self._heap[i] = self._heap[i], self._heap[parent]
            i = parent

    def _bubble_down(self, i):
        n = len(self._heap)
        while True:
            left = 2 * i + 1
            right = 2 * i + 2
            smallest = i
            if left < n and self._heap[left][0] < self._heap[smallest][0]:
                smallest = left
            if right < n and self._heap[right][0] < self._heap[smallest][0]:
                smallest = right
            if smallest == i:
                break
            self._heap[smallest], self._heap[i] = self._heap[i], self._heap[smallest]
            i = smallest


# ---------------------------------------------------------------------------
# A* path-finding
# ---------------------------------------------------------------------------

def find_path_to_target(maze: list, from_x: float, from_y: float,
                         to_x: float, to_y: float) -> list:
    """
    Find a path through the maze using A* with Manhattan heuristic.

    Parameters
    ----------
    maze   : flat Cell list from init_maze()
    from_x, from_y : start pixel coords
    to_x, to_y     : target pixel coords

    Returns
    -------
    List of Cell objects forming the path (including start), or [] if no path.
    Mirrors JS findPathToTarget().
    """
    if not maze:
        return []

    start_c = int((from_x - MAZE_OFFSET_X) // CELL_SIZE)
    start_r = int(from_y // CELL_SIZE)
    end_c = int((to_x - MAZE_OFFSET_X) // CELL_SIZE)
    end_r = int(to_y // CELL_SIZE)

    start_c = max(0, min(start_c, COLS - 1))
    start_r = max(0, min(start_r, ROWS - 1))
    end_c = max(0, min(end_c, COLS - 1))
    end_r = max(0, min(end_r, ROWS - 1))

    start = grid_index(maze, start_c, start_r)
    end = grid_index(maze, end_c, end_r)

    if start is None or end is None:
        return []

    # Reset A* state on every cell
    for cell in maze:
        cell.gCost = float('inf')
        cell.parent = None

    def heuristic(c, r):
        return abs(c - end_c) + abs(r - end_r)

    heap = MinHeap()
    start.gCost = 0
    heap.push(heuristic(start_c, start_r), start)

    while not heap.is_empty():
        _, curr = heap.pop()

        if curr is end:
            break

        if curr.gCost == float('inf'):
            continue

        for dc, dr, wall_idx in _DIRECTIONS:
            nc = curr.c + dc
            nr = curr.r + dr
            neighbor = grid_index(maze, nc, nr)

            # Skip if wall blocks passage
            if neighbor is None:
                continue
            if curr.walls[wall_idx]:
                continue
            opposite_wall = (wall_idx + 2) % 4
            if neighbor.walls[opposite_wall]:
                continue

            new_g = curr.gCost + 1
            if new_g < neighbor.gCost:
                neighbor.gCost = new_g
                neighbor.parent = curr
                f = new_g + heuristic(nc, nr)
                heap.push(f, neighbor)

    if end.gCost == float('inf'):
        return []

    # Reconstruct path
    path = []
    temp = end
    while temp is not None:
        path.append(temp)
        temp = temp.parent
    path.reverse()
    return path


# ---------------------------------------------------------------------------
# Stuck detection
# ---------------------------------------------------------------------------

def is_player_stuck(player) -> bool:
    """
    Return True if player hasn't moved more than 0.3 px since last_pos.
    Also updates last_pos to the current position for the next call.
    Mirrors JS isPlayerStuck() — note that in the original JS, last_pos is
    updated at the end of getCpuInput; here we do it in the check so the
    standalone function is self-contained for tests.
    """
    last = getattr(player, 'last_pos', None)
    if last is None:
        player.last_pos = {'x': player.x, 'y': player.y}
        return False
    dx = abs(player.x - last.get('x', player.x))
    dy = abs(player.y - last.get('y', player.y))
    stuck = dx < 0.3 and dy < 0.3
    # Update last_pos so subsequent calls reflect the latest position
    player.last_pos = {'x': player.x, 'y': player.y}
    return stuck


def get_unstuck_direction(maze: list, player=None) -> dict:
    """
    Return a direction dict {x, y} to escape when stuck.
    Prefers wall-free directions and the opposite of last movement.
    Mirrors JS getUnstuckDirection().
    """
    directions = [
        {'x': 1, 'y': 0}, {'x': -1, 'y': 0},
        {'x': 0, 'y': 1}, {'x': 0, 'y': -1},
        {'x': 1, 'y': 1}, {'x': -1, 'y': 1},
        {'x': 1, 'y': -1}, {'x': -1, 'y': -1},
    ]

    if player is None:
        random.shuffle(directions)
        return directions[0]

    check_dist = 3
    size = getattr(player, 'size', 2.0)
    cx = player.x + size / 2
    cy = player.y + size / 2

    valid = []
    for d in directions:
        check_x = cx + d['x'] * check_dist
        check_y = cy + d['y'] * check_dist
        if not is_wall(maze, check_x, check_y):
            valid.append(d)

    if not valid:
        return directions[random.randint(0, len(directions) - 1)]

    # Prefer opposite of last direction
    last_dir = getattr(player, 'last_dir', None) or {'x': 0, 'y': 0}
    ldx = last_dir.get('x', 0)
    ldy = last_dir.get('y', 0)

    if ldx != 0 or ldy != 0:
        last_dx = 1 if ldx > 0 else (-1 if ldx < 0 else 0)
        last_dy = 1 if ldy > 0 else (-1 if ldy < 0 else 0)

        scored = []
        for d in valid:
            score = 0
            if d['x'] == -last_dx and last_dx != 0:
                score += 2
            if d['y'] == -last_dy and last_dy != 0:
                score += 2
            if d['x'] != 0 and last_dx == 0:
                score += 1
            if d['y'] != 0 and last_dy == 0:
                score += 1
            score += random.random() * 0.5
            scored.append((score, d))
        scored.sort(key=lambda t: t[0], reverse=True)
        return scored[0][1]

    return valid[random.randint(0, len(valid) - 1)]
